### DISCUSSION TITLE
Memory leak in bulk upsert operations with parameter binding

### DISCUSSION THREAD

**Discussion by grzracz** [sqlalchemy/sqlalchemy#12051] [Usage Questions]
### Describe the bug

We've discovered a significant memory leak when using SQLAlchemy's parameter binding for bulk upsert operations. The memory consumption becomes unmanageable even with moderate dataset sizes, forcing us to abandon parameter binding entirely in favor of raw SQL string concatenation.

### SQLAlchemy Version in Use

2.0.36

### DBAPI (i.e. the database driver)

psycopg2

### Database Vendor and Major Version

PostgreSQL Aurora 16

### Python Version

3.12.3

### Operating system

Linux Ubuntu

### To Reproduce

```python
def upsert_with_binding(session, data_chunk):
    placeholders = []
    values = {}
    
    for idx, item in enumerate(data_chunk):
        placeholder = "(:col1_{0}, :col2_{0}, :col3_{0})".format(idx)
        placeholders.append(placeholder)
        values.update({
            f"col1_{idx}": item.val1,
            f"col2_{idx}": item.val2,
            f"col3_{idx}": item.val3,
        })
    
    sql = """
        INSERT INTO mytable (col1, col2, col3)
        VALUES {}
        ON CONFLICT (col1) DO UPDATE SET
            col2 = EXCLUDED.col2,
            col3 = EXCLUDED.col3
    """.format(",".join(placeholders))
    
    session.execute(text(sql), values)
    session.commit()
```

### Error

```
# Copy the complete stack trace and error message here, including SQL log output if applicable.
``` 

### Additional context

When performing bulk upserts with parameter binding on a table with ~13 columns, memory usage grows exponentially with the number of records. This occurs even when:

- Using chunks/batching
- Properly managing sessions with context managers
- Explicitly calling garbage collection

**Workaround**

We were forced to implement the following workaround by abandoning parameter binding entirely:

```
values_list = [f"({r.val1}, {r.val2}, {r.val3})" for r in data_chunk]
sql = f"INSERT INTO mytable VALUES {','.join(values_list)} ON CONFLICT..."
session.execute(text(sql))
```

The memory leak appears to be related to how SQLAlchemy handles the parameter dictionary for large numbers of parameters
Memory usage grows non-linearly with the number of parameters
The issue persists across multiple SQLAlchemy versions
No amount of session management or garbage collection resolves the issue
Parameter binding should maintain a reasonable memory footprint proportional to the data size, similar to raw SQL execution.

**Comment by zzzeek** [comment]
hi -

unfortunately this bug report is not actionable since it has almost no detail and there's nothing to go from.  A single function is given, with no inputs illustrated or how the function is being used, how many times it's called, etc.  There's no runnable [MCVE](http://stackoverflow.com/help/mcve) here.  How is the "leak" demonstrated?  how is the "leak" measured?     How do you know the "leak" is not in the driver?   What are the datatypes used by "mytable" ?  etc etc

please see the [mcve](https://github.com/sqlalchemy/sqlalchemy/issues?q=label%3A%22great+mcve%22+) tag for examples of how to provide an actionable example.

Also as mentioned in the issue template we prefer to start with discussions until the issue can be identified.

**Reply by zzzeek** [reply]
I would also note that a text() entry will be stored in the cache, so if you are creating enormous strings for text(), you can reduce memory use by excluding these from statement caching:

    session.execute(text(sql), values, execution_options={"compiled_cache": None})

the cache will not "leak" since it's an LRU of about 150 items, but if each SQL string is huge and unique from the others, it could use a lot of memory since `text()` will generate `BindParameter` objects for each entry.

From what this shows, it looks like you might have something that uses a lot of memory due to the `BindParameter` generation, but it's not a "leak".

For upserts, it's much more efficient to use SQLAlchemy's built in facilities for this, see https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-queryguide-upsert .   this will generate the parameter sets on the fly using a much more efficient scheme than a single giant text() object with thousands of `BindParameter` objects.
